content/blog/gotchaR-bumps.md

```diff
@@ -0,0 +1,163 @@
+use of the dollar sign, but not on a quoted variable name. Moreover, setting these options:
+```
+warnPartialMatchAttr = TRUE
+warnPartialMatchDollar = TRUE
+warnPartialMatchArgs = TRUE
+```
+in `.Rprofile` will throw a warning every time R does a partial-match. This seems like
+it might save a world of pain, although casting my mind back to various projects where I have
+frivolously talked about `list$thing` and `list$thing_name` and `list$thing_id` still makes me
+shudder a little. 
+
+Also having set these options, I occasionally see warnings popping up about auto-completion
+in code for other libraries we rely upon...
+
+## data.table vs data.frame
+
+So I collected some more data for my colleagues, and I put it in a `data.frame` like this.
+```
+data <- data.frame(stringsAsFactors = FALSE,
+  name = c("Alex", "Emma", "James", "Rich", "Rob"),
+  food = c("Oysters", "Haggis", "Goose", "Cheese", "Celery"))
+
+who_likes <- function(food) {
+  data[data$food == food, ]
+}
+```
+
+This is very nice. So if I have a haggis and wonder what to do with it:
+```
+> who_likes("Haggis")
+  name   food
+2 Emma Haggis
+```
+
+Great. However, I later want to do something astoundingly complicated with this rich dataset, and I 
+decide that using the popular [data.table](https://rdatatable.gitlab.io/data.table/) will 
+be a good idea, since it's just a higher performance version of `data.frame`, right? It's
+also in very common use. (To be clear, the performance increase is indeed very worthwhile for large
+datasets).
+
+However... if I swap in the `data.table` by only changing 
+```
+data <- data.frame(stringsAsFactors = FALSE
+```
+to
+```
+data <- data.table(stringsAsFactors = FALSE
+```
+
+then...
+
+```
+> who_likes("Haggis")
+  name    food
+1 Alex Oysters
+2 Emma  Haggis
+3 James  Goose
+4 Rich  Cheese
+5 Rob   Celery
+```
+
+Now all my friends suddenly seem to like Haggis. And Cheese too. Or do they? A bit more exploring 
+reveals that my friends in the `data.table` like absolutely everything in the universe. Even mushrooms.
+And NULL too. This may or may not be true in reality, but it is definitely not what I wanted the subset to do.
+
+What's happening is: when I write 
+`data[data$food == food, ]`, then `data.table` interprets the right-hand `food` as being
+the column in `data`, rather than the argument to `who_likes`. 
+If instead I write:-
+
+```
+who_likes <- function(x) {
+  data[data$food == x, ]
+}
+```
+
+then all works as I expected. And if I (oddly) changed `x` to `name`, then data.table
+would perform comparison between columns, returning zero rows.
+
+Looking through my previous code I have quite a number of examples where the most natural
+thing to write would be `data[data$country == country, ]` or `data[data$id == id, ]`. Luckily
+I did so only with data-frames. I have not yet found any warnings I can switch on,
+should I forget about this issue in the future. Again.
```

Comment:
I had a look to see what happens with tibbles here and got:

```
5 Rob       Celery 
> data <- tibble::tibble(data)
> who_likes <- function(food) {
+   data[data$food == food, ]
+ }
> who_likes("Goose")
# A tibble: 0 x 1
# … with 2 variables: data$name <chr>, $food <chr>
Warning messages:
1: Unknown or uninitialised column: 'food'. 
2: Length of logical index must be 1 or 5, not 0 
```

which seems somewhat reasonable